oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenExchangeAuthenticationProvider.java

```diff
@@ -0,0 +1,313 @@
+
+	/**
+	 * Constructs an {@code OAuth2TokenExchangeAuthenticationProvider} using the provided parameters.
+	 *
+	 * @param authorizationService the authorization service
+	 * @param tokenGenerator the token generator
+	 */
+	public OAuth2TokenExchangeAuthenticationProvider(OAuth2AuthorizationService authorizationService,
+			OAuth2TokenGenerator<? extends OAuth2Token> tokenGenerator) {
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		Assert.notNull(tokenGenerator, "tokenGenerator cannot be null");
+		this.authorizationService = authorizationService;
+		this.tokenGenerator = tokenGenerator;
+	}
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		OAuth2TokenExchangeAuthenticationToken tokenExchangeAuthentication =
+			(OAuth2TokenExchangeAuthenticationToken) authentication;
+
+		OAuth2ClientAuthenticationToken clientPrincipal =
+			getAuthenticatedClientElseThrowInvalidClient(tokenExchangeAuthentication);
+		RegisteredClient registeredClient = clientPrincipal.getRegisteredClient();
+
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Retrieved registered client");
+		}
+
+		if (!registeredClient.getAuthorizationGrantTypes().contains(TOKEN_EXCHANGE)) {
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.UNAUTHORIZED_CLIENT);
+		}
+
+		if (JWT_TOKEN_TYPE_VALUE.equals(tokenExchangeAuthentication.getRequestedTokenType()) &&
+			!OAuth2TokenFormat.SELF_CONTAINED.equals(registeredClient.getTokenSettings().getAccessTokenFormat())) {
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_REQUEST);
+		}
+
+		OAuth2Authorization subjectAuthorization = this.authorizationService.findByToken(
+			tokenExchangeAuthentication.getSubjectToken(), OAuth2TokenType.ACCESS_TOKEN);
+		if (subjectAuthorization == null) {
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+		}
+
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Retrieved authorization with subject token");
+		}
+
+		OAuth2Authorization.Token<OAuth2Token> subjectToken = subjectAuthorization.getToken(
+				tokenExchangeAuthentication.getSubjectToken());
+		if (!subjectToken.isActive()) {
+			// As per https://tools.ietf.org/html/rfc6749#section-5.2
+			// invalid_grant: The provided authorization grant (e.g., authorization code,
+			// resource owner credentials) or refresh token is invalid, expired, revoked [...].
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+		}
+
+		Authentication subjectPrincipal = subjectAuthorization.getAttribute(Principal.class.getName());
+		if (subjectPrincipal == null) {
+			// As per https://datatracker.ietf.org/doc/html/rfc8693#section-1.1
+			// We require a principal to be available via the subject_token for
+			// impersonation or delegation use cases.
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+		}
```

Comment:
Based on "A.1: Impersonation Token Exchange Example", the `sub` of the issued `access_token` should be the same `sub` of the `subject_token`. The way the `JwtGenerator` currently works requires a`principal` for this, so this validation ensures we have a user principal.

This would mean that `client_credentials` is not a valid grant type for the `subject_token`, because there is no user principal for that grant type. Only tokens originally issued with `authorization_code` or `device_code` would be allowed.

I wonder if there are use cases for exchanging tokens issued with the `client_credentials` grant? If so, we would have to look up the associated `RegisteredClient` and construct some kind of `Authentication` to represent that original client's principal. Seems like something we would wait to implement?

Reply:
> I wonder if there are use cases for exchanging tokens issued with the `client_credentials` grant?

Most definitely there are use cases. For example, a backend-service obtains a token using `client_credentials` grant and then calls a downstream service. The downstream service then might do a token exchange before it calls the next downstream service, but with downgraded scopes.

We need to account for this flow.